File: pinpoint/lib/model/user/user_detail_dto.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:pinpoint/model/address/address.dart';
import 'package:pinpoint/model/batch/batch_list_response.dart';

part 'user_detail_dto.g.dart';

@JsonSerializable()
class UserDetailDto {
  final String id;
  final String name;
  final String email;
  final String phone;
  final String role;
  final bool isVerified;
  final Address? address;
  final BatchListResponse? batch;
  final String? instituteId;
  final String? adminId;
  final String? adminName;
  final String createdAt;
  final String updatedAt;

  UserDetailDto({
    required this.id,
    required this.name,
    required this.email,
    required this.phone,
    required this.role,
    required this.isVerified,
    this.address,
    this.batch,
    this.instituteId,
    this.adminId,
    this.adminName,
    required this.createdAt,
    required this.updatedAt,
  });

  factory UserDetailDto.fromJson(Map<String, dynamic> json) =>
      _$UserDetailDtoFromJson(json);

  Map<String, dynamic> toJson() => _$UserDetailDtoToJson(this);
}

File: pinpointB/src/main/java/com/nxquar/pinpoint/DTO/UserDetailDto.java
package com.nxquar.pinpoint.DTO;

import com.nxquar.pinpoint.DTO.branch.BatchListResponse;
import com.nxquar.pinpoint.Model.Address;
import com.nxquar.pinpoint.Model.Users.User;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.UUID;

@Data
@AllArgsConstructor
@NoArgsConstructor
public class UserDetailDto {
    private UUID id;
    private String name;
    private String email;
    private String phone;
    private String role;
    private boolean isVerified;
    private Address address;
    private BatchListResponse batch;
    private UUID instituteId;
    private UUID adminId;
    private String adminName;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    public static UserDetailDto fromEntity(User user) {
        return new UserDetailDto(
                user.getId(),
                user.getName(),
                user.getEmail(),
                user.getPhone(),
                user.getRole().toString(),
                user.isVerified(),
                user.getAddress(),
                user.getBatch() != null
                        ? new BatchListResponse(
                        user.getBatch().getId(),
                        user.getBatch().getName(),
                        user.getBatch().getCode()
                )
                        : null,
                user.getInstitute() != null ? user.getInstitute().getId() : null,
                user.getAdmin() != null ? user.getAdmin().getId() : null,
                user.getAdmin() != null ? user.getAdmin().getName() : null,
                user.getCreatedAt(),
                user.getUpdatedAt()
        );
    }

}
